Products/ZPerFactMods/requestWrapper.py

```diff
@@ -0,0 +1,89 @@
+#
+# Hook into event handlers to call request_init/request_end
+#
+
+import zope.component
+import ZPublisher.pubevents
+import six
+import logging
+
+logger = logging.getLogger('Products.ZPerFactMods.requestWrapper')
+
+
+def get_context(event):
+    """
+    Look up the context of the published object. Returns None for some
+    irregular requests, for example for builtin ZMI assets that have no title
+    or when publishing a File object.
+    Also returns None if the published object's title contains NO_REQWRAP.
+    If the context does not contain all necessary acquisition components as
+    defined by the nav_acquire property of the site object returned by
+    get_site(), returns the site object instead.
+    """
+    try:
+        title = event.request.PUBLISHED.title
+        if callable(title):
+            title = title()
+        if isinstance(title, six.binary_type) and b'NO_REQWRAP' in title:
+            return None
+        if isinstance(title, six.text_type) and u'NO_REQWRAP' in title:
+            return None
+
+        context = event.request.PARENTS[0]
+        parents = {obj.id for obj in event.request.PARENTS}
+        root = context.get_site()
+        nav_acquire = set(root.nav_acquire)
+
+        if len(nav_acquire.difference(parents)):
+            # Parts from nav_acquire are missing
+            return root
+        return context
+
+    except Exception:
+        # Hook not implemented, published object has no title, ...
```

Comment:
Or, an error in `get_site()`. Or, no `get_site()` implemented, or, no `nav_acquire` settings. Quite a few ways this can go wrong…

Maybe a log line containing the original exception would be helpful if people are unlucky enough to use a less-than-perfect installation.

Reply:
I originally tried to log every caught exception, but this is fired all the time for both internal ZMI methods and also when publishing `File` objects.
However: I think I can split this. Check if there is a `title` attribute and return `None` otherwise without comment. Then wrap the construction of the context in a `try .. except` which creates a log line if something happens. I think there were also cases where `event.request.PARENTS[0]` yielded an error, but I will check this out again.